### PR TITLE
Mobile UI for desktop

### LINKED ISSUE #7399: Use same UI and navigation on desktop and mobile
# Problem

Desktop is currently using a different set of screens and navigation system, which adds up to the burden of maintaining the platform. 

# Implementation

Check what is required to have desktop work with the current mobile screens. We can use conditionals like we already have for ios/android for platform specific things but overall that would remove the only limitation we have for having features released on both platforms simultaneously.

# Acceptance Criteria

Desktop doesn't use it's own screens anymore

# Related issues
- [Improve FlatList performance ](https://github.com/status-im/react-native-desktop/issues/470)
- [Two-pane status UI](https://github.com/status-im/status-react/issues/8168)

# Notes

- react-navigation might need to be implemented for the desktop platform?

### PR DESCRIPTION
**Pull request by vkjr**
fixes #7399

### Summary:

Mobile UI used for desktop platform

### Review notes (optional):
Most issues solved on `react-native-desktop` side and with the help of `(if desktop?...)`.
But there is one area that needs special attention during review:
On mobile `basic-text-input` component saves entered text in re-frame database. Whenever input changes, component recreates. That doesn't cause issues because typing speed on mobile is much slower.
But on desktop typing speed significantly faster and re-rendering causes characters loss. So I had to create component `basic-text-input-desktop` and save input text inside component state instead of re-frame database. Drawback of that decision - typed text doesn't save when I close chat and reopen it again. 
Would appreciate your help in solving this.

### Testing notes (optional):
I believe mobile builds should be tested especially careful because problems in desktop can be solved in a separate issues, but we don't want any regressions on mobile.

#### Platforms (optional)
- Android
- iOS
- macOS
- Linux
- Windows

#### Areas that maybe impacted (optional)
**Functional**
- 1-1 chats
- public chats
- group chats
- wallet / transactions
- dapps / app browsing
- account recovery
- new account
- user profile updates
- networks
- mailservers
- fleet
- bootnodes

**Non-functional**
- battery performance
- CPU performance / speed of the app
- network consumption

### Steps to test:
- Open Status
- check everything..

status: ready

### CODE REVIEW

**Comment by cammellos** [comment]
@vkjr for the issue with the input text, I think storing it in local storage is a better strategy, as you get the same issue on mobile if using a keyboard for example. To preserve the current behavior (close / open chat and text is stored), we could store the text on the component unmounting  `componentWillUnmount` and store the chat text in the db, and conversely load it on `componentDidMount`.

File: src/status_im/ui/screens/chat/input/input.cljs [hunk 1]
```diff
@@ -51,6 +52,45 @@
       (when cooldown-enabled?
         {:placeholder (i18n/label :cooldown/text-input-disabled)}))]))
 
+(defview basic-text-input-desktop [{:keys [set-container-width-fn height single-line-input? set-text state-text]}]
+  (letsubs [cooldown-enabled?    [:chats/cooldown-enabled?]]
+    ;(let [component               (reagent/current-component)
```
Comment (cammellos) [inline src/status_im/ui/screens/chat/input/input.cljs]: Probably worth removing if not needed

File: src/status_im/ui/screens/chat/input/input.cljs [hunk 2]
```diff
@@ -135,7 +180,13 @@
             {:keys [input-text]} [:chats/current-chat]
             result-box           [:chats/current-chat-ui-prop :result-box]
             show-stickers?       [:chats/current-chat-ui-prop :show-stickers?]]
-    (let [single-line-input? (:singleLineInput result-box)]
+    (let [single-line-input? (:singleLineInput result-box)
+          component               (reagent/current-component)
```
Comment (cammellos) [inline src/status_im/ui/screens/chat/input/input.cljs]: indentation looks a bit odd

**Review by cammellos — APPROVED** [review]

**Comment by vkjr** [comment]
@cammellos, thanks, great idea about `componentWillUnmount`!

File: src/status_im/desktop/core.cljs [hunk 3]
```diff
@@ -1,25 +1,37 @@
 (ns status-im.desktop.core
   (:require [reagent.core :as reagent]
-            [re-frame.core :as re-frame]
+            [re-frame.core :refer [subscribe dispatch dispatch-sync]]
```
Comment (flexsurfer) [inline src/status_im/desktop/core.cljs]: why did you decide to use `refer` instead alias?

File: src/status_im/desktop/core.cljs [hunk 4]
```diff
@@ -1,25 +1,37 @@
...
+      (fn []
+        (.stop react/http-bridge)
+        (.removeEventListener react/app-state "change" app-state-change-handler))
+      :display-name "root"
+      :reagent-render views/main})))
 
 (defn init []
-  (core/init app-root))
+  (core/init app-root)
+  ;(snoopy/subscribe!)
```
Comment (flexsurfer) [inline src/status_im/desktop/core.cljs]: remove?

File: src/status_im/desktop/core.cljs [hunk 5]
```diff
@@ -1,25 +1,37 @@
...
 
 (defn app-root [props]
-  (reagent/create-class
-   {:component-did-mount (fn [this]
-                           (re-frame/dispatch [:set-initial-props (reagent/props this)])
-                           (shortcuts/register-default-shortcuts)
-                           (deep-links/add-event-listener))
-    :reagent-render      (fn [props]
-                           views/main)}))
+  (let [keyboard-height (subscribe [:get :keyboard-height])]
```
Comment (flexsurfer) [inline src/status_im/desktop/core.cljs]: why do we need keyboard on desktop?

File: src/status_im/ui/components/list_selection.cljs [hunk 6]
```diff
@@ -21,9 +22,12 @@
     :action #(open-share {:message text})}])
 
 (defn show [options]
-  (if platform/ios?
-    (action-sheet/show options)
-    (dialog/show options)))
+  (when platform/ios?
```
Comment (flexsurfer) [inline src/status_im/ui/components/list_selection.cljs]: better to use http://clojuredocs.org/clojure.core/cond

**Review by flexsurfer — APPROVED** [review]

**Comment by flexsurfer** [comment]
Great job!

**Comment by cammellos** [comment]
Great job, taking for a spin, in the meantime: 

Issue: Search box has the same issue as input box (keypress not registered, likely is using the re-frame db)

**Comment by churik** [comment]
@vkjr thanks for a lot of work!

1) The scroll is almost unusable in chat view when history  is quite long (~50+ messages)
It is jumped all the time and I can't easily scroll from today to yesterday for example.
On [this video](https://take.ms/E7cMW) comparison between currebt desktop version <-> this PR when I try to scroll through chat history in Status channel

2) I expected that wallet and DApps also will be added in this version. 
If not, we should think what to do with /send, /request and ENS names app and with extensions/stickers and also with "Send Transaction" in profile screen

3) No way to open user profile from public chat

4) No way to open Search in chats (mentioned in https://github.com/status-im/status-react/pull/7551#issuecomment-474310271)

5) Two last sent messages are overlapped 
image

6) Messages in chat list are cut when they have > 1 row
выавыаыва

7) Can't copy \ reply to a message - no context menu is shown 

8) Nothing happens when perform some actions (list is below)
 8.1 Tap on any chat > three dots > Share chat - nothing happens
 8.2 + > Invite friends
 8.3 Contacts > open any contact > Share profile link
 8.4 Profile > Share my profile > Share link
 8.5 New group chat > 

9) Error at attempt to set photo from gallery or to capture it
asdasds

10) Crash at attempt to interact with browser window when tap on "ENS names"
Steps:
- open profile
- tap on ENS names
- tap on reload
Crash logs: [Archive.zip](https://github.com/status-im/status-react/files/2983556/Archive.zip)

11) Click on Profile > Notifications shows error "There is no application set.."
sdf
I actually doubt that we need this section and Mobile data as well (it is mobile specific I believe)

12) Question: StatusIm.log is not created anymore, only geth.log?

13) I can add an extension, but it doesn't work (i.e. added gfycat from https://status.im/extensions/examples.html) and still see `unhandled command` in message instead of gif

14) Can't connect to another network 
When I tap on "Connect" I see "Connecting to another network requires login" - no logout as in mobile
connecting

15) No PNs are received on desktop

16) Scan QR in Add Bootnode, Add extension, New > Scan QR shows an error
It's definitely only mobile feature
QR

Tested on Mac OSx 10.14.3

**Comment by oskarth** [comment]
@vkjr This PR has been open for a long term. Can we please merge partial progress under a flag and do smaller PRs in the future with tighter test scope (https://github.com/status-im/status-react/blob/develop/doc/decisions/0011-tweak-pr-process.md)

**Comment by vkjr** [comment]
@oskarth, good idea about merging partial progress under a flag, thanks!
There is only one significant problem left - `FlatList` performance. If it solved, we can test/merge the PR without additional flags. If I won't find the problem during 2 days, I will add a flag and we will merge PR as is. Does it sound suitable?

**Comment by oskarth** [comment]
@vkjr sounds good, thanks!

Out of curiosity, how bad is the perf for flatlist?

**Comment by mandrigin** [comment]
@vkjr if that is behind the flag, let's merge it. I don't think it should be default though until we have a proper 2-pane UI.

**Comment by vkjr** [comment]
@oskarth very :( Almost unusable on a 50+ messages.
In previous version of desktop UI we used a non-standard version of ScrollView, specifically optimized on Qt side. But now it would be better to fix the `FlatList` itself, to avoid difference between mobile and desktop code.

**Comment by oskarth** [comment]
Ok, thanks.

Happy birthday btw!

**Comment by vkjr** [comment]
@oskarth, thank you! :)
@mandrigin, ok, I'll add a flag! Will continue with FlatList performance in parallel.

**Comment by vkjr** [comment]
Now mobile UI launches for desktop only if in `.env` file specified `MOBILE_UI_FOR_DESKTOP=1`. Otherwise old (current) desktop ui used.

@churik, could you please take a look at this PR at another angle? We need to make sure that it doesn't break any existing functionality in old desktop UI and mobile builds. If everything ok, we can merge it and continue work on mobile ui for desktop in future, smaller PRs.

**Comment by jakubgs** [comment]
>We need to make sure that it doesn't break any existing functionality in old desktop UI and mobile builds. If everything ok, we can merge it and continue work on mobile ui for desktop in future, smaller PRs.

I totally agree. If this doesn't break our legacy desktop UI then let's merge it while it has a manageable size. Otherwise this will keep growing and growing until only Volodymyr knows what is happening here, or worse, even he doesn't anymore.

**Comment by churik** [comment]
@vkjr 
**1) No names for context menus on desktop builds**
dsdoifsdifijofsd
Error at attempt to use any of items in context menu (i.e. copy message, clear history, ect)
dsfsfsdfsddsf

**2) Error on start android e2e build**
See resuts here: 
klfdkslklflll
[logcat.log](https://github.com/status-im/status-react/files/3095138/logcat.log)

**3) Can't build Android**
```
17:23:01  > Task :app:generatePrBuildConfig
17:23:14  
17:23:14  > Task :app:bundlePrJsAndAssets
17:23:14  warning: the transform cache was reset.
17:23:16  jest-haste-map: Watchman crawl failed. Retrying once with node crawler.
17:23:16    Usually this happens when watchman isn't running. Create an empty `.watchmanconfig` file in your project's root folder or initialize a git or hg repository in your project.
17:23:16    Error: Watchman error: A non-recoverable condition has triggered.  Watchman needs your help!
17:23:16  The triggering condition was at timestamp=1555600994: inotify-add-watch(/home/jenkins/workspace/status-react_prs_windows_PR-7999/node_modules/babel-preset-react-native/node_modules/@babel/helper-module-transforms) -> The user limit on the total number of inotify watches was reached; increase the fs.inotify.max_user_watches sysctl
17:23:16  All requests will continue to fail with this message until you resolve
17:23:16  the underlying problem.  You will find more information on fixing this at
17:23:16  https://facebook.github.io/watchman/docs/troubleshooting.html#poison-inotify-add-watch. Make sure watchman is running for this project. See https://facebook.github.io/watchman/docs/troubleshooting.html.
17:23:21  events.js:167
17:23:21        throw er; // Unhandled 'error' event
17:23:21        ^
```

**4) Please, check packeges-check-bot: https://github.com/status-im/status-react/pull/7551/checks?check_run_id=104980620**

**Comment by vkjr** [comment]
@jakubgs, could you please take a look at android build? Getting `Error: Watchman error: A non-recoverable condition has triggered.  Watchman needs your help!` issue.

**Comment by jakubgs** [comment]
Okay, I've increased `max_user_watches` from `65536` to `131072`: https://github.com/status-im/infra-ci/commit/3033ce38cfd7bf6b9a62e0fca137933d98963232
```
 $ ansible ci-slave-linux -o -a 'cat /proc/sys/fs/inotify/max_user_watches'
linux-01.do-ams3.ci.misc | CHANGED | rc=0 | (stdout) 131072
linux-03.do-ams3.ci.misc | CHANGED | rc=0 | (stdout) 131072
linux-05.do-ams3.ci.misc | CHANGED | rc=0 | (stdout) 131072
linux-04.do-ams3.ci.misc | CHANGED | rc=0 | (stdout) 131072
linux-02.do-ams3.ci.misc | CHANGED | rc=0 | (stdout) 131072
```
Try again.

**Comment by jakubgs** [comment]
Looks like you branch has conflicts. Please rebase.

**Comment by churik** [comment]
@vkjr 
For mobile builds:
1) 'Error: Error: Error: Error: Error: Unknown icon:' 
- tapping on + > start new chat
- tap on profile icon in public chat
2) No Dapps tab on Android

For Desktop builds:
3) Tappable area is changed for user profile icon  in public chat (desktop)
sdfsdfsdfdsfsdf
For now it works only if you tap on the top of the icon.
On current nightly  whole icon is clickable

**Comment by churik** [comment]
I don't know why it is moved to `to test` automatically, but needs @vkjr attention

**Comment by vkjr** [comment]
@churik, thanks for checking!

**Comment by 0xc1c4da** [comment]
➤ Oskar T commented:

Copying update:

Hi!
As for mobile UI. Tetiana tested latest changes yesterday and found 3 issues, one per platform) Working on them today. Yes, I think it is possible to merge it this week.

As for FlatList, its behavior consist of few different issues. First of them - long delays before loading next chunk of data. For this issue solution was found yesterday. There is discrepancy between mobile and desktop events flow. Desktop sends only one onScroll event per scroll so doesn't trigger js implementation code as many times as it should.

Second issue - FlatList scroll position jumps when elements dynamically added to while it is moving. Will take this right after fixes needed for mobile UI.

**Comment by vkjr** [comment]
@churik, problems you mentioned fixed.

Issue with desktop happened due to fixes in `react-native-desktop` for mobile UI. For now I reverted them. So we have old desktop UI working fine with correct click area for avatar. But new mobile UI again doesn't open profile from chat. Lets merge this code under a flag and I will fix it in a separate issue.

**Comment by churik** [comment]
@vkjr please, fix conflict and I'll try hopefully last time)

**Comment by vkjr** [comment]
@churik, fixed

**Comment by churik** [comment]
@vkjr 
some regression:
#### **3) Input value is not preserved in the message field (IOS, Android)**
**Steps:**
- create account
- join public channel
- type smth in the input field
- tap "Back"
- open this public chat again
**Expected result:** input is preserved
**Actual result:** input is erased

#### **4) No background for "Edit" when editing profile photo (IOS, Android)**

photo_2019-04-29 09 47 19 jpeg 2019-04-29 09-49-35

**Comment by vkjr** [comment]
@churik, issues were fixed

**Comment by churik** [comment]
Issues are fixed.
However, I have some concerns about this PR.

1) I am not quite sure that I found all the issues that can be introduced by this PR.
Taking into account https://github.com/status-im/status-react/pull/7551#issuecomment-487481833, https://github.com/status-im/status-react/pull/7551#issuecomment-485808018 and https://github.com/status-im/status-react/pull/7551#issuecomment-484571672 issues can appear in any area of app - desktop or mobile.
I did exploratory testing of all 5 builds during several rounds of testing, and as the main goal of Status remains mobile app I would suggest whole regression test scope for this PR on Android and IOS at least (as we do for release)

2) As decided from https://github.com/status-im/status-react/pull/7551#issuecomment-483982088 mobile UI will be hidden for now because we need to polish it and have a version that is equal or better than the current Desktop app.
Meanwhile, we are doing a lot of work on refactoring/adding new features/redesign, etc and there is no chance to track do these changes affect mobile UI or not.
How can we align new PRs with Mobile UI if it is hidden? should developers\QA check Mobile UI for new PRs?

Appreciate comments on both points.
cc @oskarth @annadanchenko @mandrigin

**Comment by mandrigin** [comment]
I wouldn't worry that much about (2) though, the whole point of having the same UI on Desktop and on mobile is to share new features and refactorings. That will help us to double check if that is a viable approach without going all-in there.

As of (1), well, I guess the next release testing will show everything else that regressed :-D

**Comment by yenda** [comment]
@churik I agree with @mandrigin regarding regressions we just had a release cut which leaves lot of time to find potential regressions so I wouldn't worry too much about that. To make (2) happen faster I would completely give up on implementing anything for desktop for now on.

**Comment by vkjr** [comment]
@mandrigin, can I merge this?
@churik, it still has `Tested-Issues` tag

**Comment by mandrigin** [comment]
Let’s wait for @churik to confirm first. After that — we can merge.

**Comment by annadanchenko** [comment]
@mandrigin @yenda @oskarth @vkjr 
(1) I'd rather do full regression of mobile app before merging this PR as we can't just close the eyes and say ok we have a week to fix any possible side effects on mobile app. It looks from @churik comments that there were issues in unexpected areas and more can be hiding, so breaking stability of nightlies builds and new PRs.
(2) the challenge is that after merging this PR behind the flag we will not be able to test that any other PR works well and does not introduce issues for mobile ui (if any). So, we need to understand that later more issues will arise and existing features implemented in mobile will have issues on mobile desktop ui. Even to validate the assumption that such thing will not happen we will need a special build to "uncover" the desktop mobile ui. Shall we agree on the process here on what would be the miletstones for mobile ui development, @oskarth @mandrigin ?

**Comment by vkjr** [comment]
Regarding topic (2), I think currently we can treat this PR not as a "merging working mobile UI for desktop", but as a "preparing code for new UI". So I'd ignore testing of new UI with other PRs after merge. Instead full testing can be made when new UI will become default.

**Comment by annadanchenko** [comment]
> Regarding topic (2), I think currently we can treat this PR not as a "merging working mobile UI for desktop", but as a "preparing code for new UI". So I'd ignore testing of new UI with other PRs after merge. Instead full testing can be made when new UI will become default.

sure, I'd expect full testing as soon as mobile ui is fully ready. The question here if there are any extra milestones we'd like to include (e.g. after new onboarding is implemented in mobile app -> enable mobile desktop ui and see how does it look there)

**Comment by oskarth** [comment]
I agree with what @mandrigin said above. The sooner we can get this into the code base under a flag the better. If adequate testing has been done on this PR I think that's fine.

For dependencies to actually enable it, my understanding is that this mostly about FlatList performance. If there are other issues, can we please capture these in some umbrella issue? Perhaps there is one already?

**Comment by annadanchenko** [comment]
adequate testing for this PR needs regression testing for mobile app (at least executing Critical/High test cases) that @churik will do today

**Comment by vkjr** [comment]
@oskarth, I considered #7399 as an "umbrella issue". Currently FlatList is an only dependency from what I see. It is listed in section "related issues".

**Comment by churik** [comment]
@vkjr please, rebase to current develop and resolve conflicts one more time.
Thanks!

**Comment by vkjr** [comment]
Please, do not test latest commit with rebase. I found typing issue in old desktop UI, fixing it now.

**Comment by vkjr** [comment]
@churik, I fixed the issue, can be tested now

**Comment by churik** [comment]
@vkjr still testing, so far tested on Linux Ubuntu 18.04, Windows 10 Home, Mac OSx:

- fresh install, account creating
- upgrade from current nightly
- compatibility with 0.12.0
- exploratory testing in chats, PNs and Profile

What I noticed that clicks on user icons in public chats are still not working as expected, so sometimes it does nothing.
See video: https://take.ms/zS2su
Click on user icon works for me in ~20% cases.
Log: [Status_click.log](https://github.com/status-im/status-react/files/3138694/Status_click.log)


Continue with testing on mobile...

**Comment by vkjr** [comment]
@churik, my bad, used incorrect `react-native-desktop` version. Fixed.

**Comment by churik** [comment]
@vkjr test-list for mobile: https://ethstatus.testrail.net/index.php?/runs/view/3246&group_by=cases:section_id&group_order=asc

Still testing.
With last commit issues from https://github.com/status-im/status-react/pull/7551#issuecomment-487481833 are again reproducible (checked on IOS)
Please take a look.

**Comment by vkjr** [comment]
@churik, should be fine now

**Comment by churik** [comment]
@vkjr what regression can cause last commit?
Asking because I've already tested most of the TC.
And as you squashed commits, I can identify what exactly changes are done since https://github.com/status-im/status-react/pull/7551#issuecomment-489074414

**Comment by vkjr** [comment]
@churik, last commit shouldn't cause any. It appeared that by one of the latest commit I overwrote the one with ios\android fixes. So in a last one only those fixes were returned back.

**Comment by vkjr** [comment]
As for squishing commits - won't do like this anymore, my hopes for quick merge were too optimistic :)

**Comment by churik** [comment]
Tested on desktop builds: https://github.com/status-im/status-react/pull/7551#issuecomment-488713451
Rechecked issue from https://github.com/status-im/status-react/pull/7551#issuecomment-488713451 on Mac OSx.
Test list for mobile is [here](https://ethstatus.testrail.net/index.php?/runs/view/3246&group_by=cases:section_id&group_order=asc)
(most of test-cases are checked on build with this commit: https://github.com/status-im/status-react/commit/cfb94272fb2aff36d3db0f699a2b9f84376f959d)
Primary device: IOS 11.4.1 (IPhone 7), secondary device: Android 8 (LG V20)

No regression introduced by this PR is found. All failures mentioned in the test suite are failures in nightly as well.

**Pull request merged**